rpc_codegen/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+        let service_name = &service.name;
+        let name = format!("{}WithContext", service_name);
+        let trait_name: TokenStream = name.parse().unwrap();
+
+        let mut methods: Vec<TokenStream> = vec![];
+        for method in &service.methods {
+            methods.push(match (method.client_streaming, method.server_streaming) {
+                (false, false) => self.generate_add_unary_call(&method),
+                (false, true) => self.generate_add_server_streams_procedure(&method),
+                (true, false) => self.generate_add_client_streams_procedure(&method),
+                (true, true) => self.generate_add_bidir_streams_procedure(&method),
+            });
+        }
+        quote! {
+        pub fn register_service<
+                S: #trait_name<Context> + Send + Sync + 'static,
+                Context: Send + Sync + 'static
+            >(
+                port: &mut RpcServerPort<Context>,
+                service: S
+            ) {
+                let mut service_def = ServiceModuleDefinition::new();
+                // Share service ownership
+                let shareable_service = Arc::new(service);
+
+                #(#methods)*
+
+                port.register_module(#service_name.to_string(), service_def)
+            }
+        }
+    }
+
+    fn generate_add_unary_call(&self, method: &Method) -> TokenStream {
+        let method_name: TokenStream = method.name.parse().unwrap();
+        let proto_method_name = &method.proto_name;
+        let input_type: TokenStream = method.input_type.parse().unwrap();
+        quote! {
+            let service = Arc::clone(&shareable_service);
+            service_def.add_unary(#proto_method_name, move |request, context| {
+                let service = service.clone();
+                Box::pin(async move {
+                    let response = service
+                        .#method_name(#input_type::decode(request.as_slice()).unwrap(), context)
+                        .await;
+                    response.encode_to_vec()
+                })
+            });
+        }
+    }
+
+    fn generate_add_server_streams_procedure(&self, method: &Method) -> TokenStream {
+        let method_name: TokenStream = method.name.parse().unwrap();
+        let proto_method_name = &method.proto_name;
+        let input_type: TokenStream = method.input_type.parse().unwrap();
+        quote! {
+            let service = Arc::clone(&shareable_service);
+            service_def.add_server_streams(#proto_method_name, move |request, context| {
+                let service = service.clone();
+                Box::pin(async move {
+                    let server_streams = service
+                        .#method_name(#input_type::decode(request.as_slice()).unwrap(), context)
+                        .await;
+                    // Transforming and filling the new generator is spawned so the response it quick
```

Comment:
Did you intend to write?
```suggestion
                    // Transforming and filling the new generator is spawned so the response is quick
```